library/Fission/Internal/UTF8.hs

```diff
@@ -18,6 +18,21 @@ import qualified RIO.ByteString      as Strict
 import qualified RIO.ByteString.Lazy as Lazy
 import qualified RIO.Text            as Text
 
+
+-- Property testing
+--
+-- $setup
+-- >>> :set -XOverloadedStrings
+-- >>> import Test.QuickCheck
+-- >>> import Test.QuickCheck.Instances ()
+-- >>> import qualified Data.Text as Text
+-- >>> import qualified RIO.ByteString.Lazy as Lazy
+
+
+
+-- Instances
```

Comment:
Soooo two things:

1. `class` is a class, not an instance
2. Do we need to announce every section? To me this needlessly wordy to me, akin to:

```haskell
do
  emailUser user -- emails user
  updateDB emailed True -- update DB that user was emailed
```

Given that we have syntax to announce that these are classes, instances, and functions, why not just:

```haskell
-- Property testing
--
-- $setup
-- >>> :set -XOverloadedStrings
-- >>> import Test.QuickCheck
-- >>> import Test.QuickCheck.Instances ()
-- >>> import qualified Data.Text as Text
-- >>> import qualified RIO.ByteString.Lazy as Lazy

class Textable a where
  encode :: a -> Either UnicodeException Text
  encode :: a -> Either UnicodeException Text

instance Textable ByteString where
  encode = decodeUtf8'

instance Textable Lazy.ByteString where
  encode = encode . Lazy.toStrict

showLazyBS :: Show a => a -> Lazy.ByteString
showLazyBS = textToLazyBS . textDisplay . displayShow
```

...or break the the class & instances into their own module?

Thoughts?

Reply:
So, I like to divvy things up that feel separate, that's why I leave so much space between those things. I like how I have a better visual overview when I do this.

That said, it doesn't make sense to me to create new files for stuff that's so small. I only like to separate things into multiple files if one file is over a 1000 lines long or more.

It's a style I adopted from Elm.
I can use your style if this bothers you?

Reply:
>  it doesn't make sense to me to create new files for stuff that's so small

It's small now, but can grow over time.

> It's a style I adopted from Elm.

Haskell and Elm have different features. Your right that in Elm you absolutely need comments to break apart the sections of your code, but that's a consequence of Elm having fewer options to structure your application. You have a bunch of `map` and `andThen` pseudo-instances, for example, and probably want those all labelled as such. Haskell has more options for abstraction, and so also more keywords like `instance`, so we don't need another label announcing that it's an instance.

> I only like to separate things into multiple files if one file is over a 1000 lines long or more.

Again, very different trade offs between languages. Purely on style 1000 LOC seems pretty steep to me in any language — a file over _200 lines_ is acceptable but should be rare.

For broken down modules, it depends on if you want consistency or not. If we have all of our stuff broken into `*.Types`, it makes more sense than sometimes having them and sometimes not. Think of this as being cut the other way like in Phoenix: you have a directory called "controllers" where all of your short controllers live, and another called "views" where your short views live. We're slicing the stack the other way, from domain down.

Splitting out `.Types` is important because it helps prevent circular dependencies. Splitting out `.Class` for the `class` and instances for pre-existing types helps get rid of a bunch of noise in the module.

Reply:
> Yes please 

You got it, will change 👍 

> It's small now, but can grow over time.

True, that said, I'm more of the "split it up when needed" kind of guy.

> Purely on style 1000 LOC seems pretty steep to me in any language

Yeah, I get that a lot. Doesn't bother me much really 🤷‍♂ 
My reasoning is more like:
- Say you would have a `Text` module and everything related to text is in that module
- I would just use my editor to navigate between all the functions and types. I can easily browse through all the functions and types using autocomplete.
- If things are in separate files, I have to do more effort to look for things.

But yeah, totally personal preferences.
Will change 😉

Reply:
@icidasset thanks! And extra thanks for explaining your reasoning and engaging in the conversation in a strong-opinions-loosely-held way 🎉🎉🎉


![](https://media1.giphy.com/media/xULW8v7LtZrgcaGvC0/giphy.gif?cid=36b14fac844f95a4f3fa180594bfb8cb76f0066322c5fea9&rid=giphy.gif)